application/src/main/java/io/smallrye/context/application/context/propagation/ApplicationContextProvider.java

```diff
@@ -0,0 +1,47 @@
+package io.smallrye.context.application.context.propagation;
+
+import java.util.Map;
+
+import org.eclipse.microprofile.context.ThreadContext;
+import org.eclipse.microprofile.context.spi.ThreadContextProvider;
+import org.eclipse.microprofile.context.spi.ThreadContextSnapshot;
+
+public class ApplicationContextProvider implements ThreadContextProvider {
+
+    @Override
+    public ThreadContextSnapshot currentContext(Map<String, String> props) {
+        ClassLoader capturedTCCL = Thread.currentThread().getContextClassLoader();
+        System.err.println("CAPTURING TCCL " + capturedTCCL);
+        return () -> {
+            ClassLoader movedTCCL = Thread.currentThread().getContextClassLoader();
+            if (capturedTCCL != movedTCCL)
+                Thread.currentThread().setContextClassLoader(capturedTCCL);
+            return () -> {
+                ClassLoader latestTCCL = Thread.currentThread().getContextClassLoader();
+                if (latestTCCL != movedTCCL)
+                    Thread.currentThread().setContextClassLoader(movedTCCL);
+            };
+        };
+    }
+
+    @Override
+    public ThreadContextSnapshot clearedContext(Map<String, String> props) {
```

Comment:
Does this even make sense? Setting TCCL to `null` can have unwanted side effects. Does the spec require this? IMO it makes more sense to just "not touch it" is case of `clearedContext`, WDYT?

Reply:
Well, I guess that's a question for the spec. We have "cleared" as opposed to "unchanged" so…

Reply:
I am aware of that, but I can't imagine anyone actually clearing TCCL like this...
I will ask this on the spec issue.

Reply:
Asked at https://github.com/eclipse/microprofile-context-propagation/pull/194#issuecomment-660928422

Reply:
Saw that, we both did apparently :-)